src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -116,8 +116,9 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
             }
 
             var documentVersionStamp = await documentSnapshot.GetTextVersionAsync();
```

Comment:
Sounds like it may be time to add a "Version" to documentSnapshot instead of relying on the text version + project version. You can verify with Roslyn if this is how their text documents also function ( I'd be shocked if they didn't ). Ultimately my beef is that we shouldn't need to look at the document + project version to understand that important state has changed, a document version should reflect that. You can kind of get that with the "generated C# output version" today but that only "kind of" works because if the C# doesn't actually change then you don't rev appropriately.

Reply:
Seems like [this](https://github.com/dotnet/roslyn/blob/6ce0ed7dd542240025a6eae11f0028f526246049/src/Workspaces/Core/Portable/Workspace/Solution/Document.cs#L125-L130) is their equivalent. Will consult with them though.